src/utils/key.rs

```diff
@@ -51,43 +70,56 @@ impl KeyBundle {
         let key_bits = self.bits;
         let priv_key = match self.key_type.as_str() {
             "rsa" => {
-                if key_bits != 2048 && key_bits != 3072 && key_bits != 4096 {
-                    return Err(RvError::ErrPkiKeyBitsInvalid);
+                match key_bits {
+                    2048 | 3072 | 4096 => {
+                        let rsa_key = Rsa::generate(key_bits)?;
+                        PKey::from_rsa(rsa_key)?.private_key_to_pem_pkcs8()?
+                    },
+                    _ => return Err(RvError::ErrPkiKeyBitsInvalid),
                 }
-                let rsa_key = Rsa::generate(key_bits)?;
-                let pkey = PKey::from_rsa(rsa_key)?;
-                pkey.private_key_to_pem_pkcs8()?
             }
             "ec" => {
                 let curve_name = match key_bits {
                     224 => Nid::SECP224R1,
                     256 => Nid::SECP256K1,
                     384 => Nid::SECP384R1,
                     521 => Nid::SECP521R1,
-                    _ => {
-                        return Err(RvError::ErrPkiKeyBitsInvalid);
-                    }
+                    _ => return Err(RvError::ErrPkiKeyBitsInvalid),
                 };
                 let ec_group = EcGroup::from_curve_name(curve_name)?;
-                let ec_key = EcKey::generate(ec_group.as_ref())?;
-                let pkey = PKey::from_ec_key(ec_key)?;
-                pkey.private_key_to_pem_pkcs8()?
-            }
-            "aes-gcm" | "aes-cbc" | "aes-ecb" => {
-                if key_bits != 128 && key_bits != 192 && key_bits != 256 {
-                    return Err(RvError::ErrPkiKeyBitsInvalid);
+                let ec_key = EcKey::generate(&ec_group)?;
+                PKey::from_ec_key(ec_key)?.private_key_to_pem_pkcs8()?
+            },
+            "sm2" => {
```

Comment:
It seems this needs `#[cfg(not(tongsuo))]` as well?

Reply:
Unless you may compile this code only with OpenSSL 1.1.1 or later. By luck that users are not using an older version of OpenSSL.

Reply:
Yes, fixed.